### ISSUE TITLE
Wire or deprecate CompiledClassMetadataFactory & co.

### ISSUE THREAD

**Issue by nicolas-grekas** [symfony/symfony#57552]
CompiledClassMetadataFactory was added in #29117 by @fbourigault, but it's not used anywhere in the codebase.

I'm wondering if we shouldn't deprecate it (and it's related classes, like the cache warmer) or if we should figure out a way to wire it (or if it's used somewhere else?)

Anyone would like to dig the topic?

**Comment by fbourigault** [comment]
Since this development, there are other approaches about serializer performance ([Un Serializer sous stéroïdes](https://live.symfony.com/2024-paris/schedule/serialisation-et-performance)).

@Korbeil, @mtarld does this make sense as Automapper CacheWarmer can achieve a greater performance boost for anyone requiring it?

**Comment by Korbeil** [comment]
Hey ! sorry for the delay, I had the issue pinned but forgot to answer :sob: 

I'm not familiar with the feature you introduced but it seems it will compile the collected metadata to make metadata fetching faster the second time you call it. It's a different approach then AutoMapper (or JsonEncoder) where we do fetch metadata once then we generate the transformer (Normalizer equivalent) so we don't need the metadata anymore.

I do think your compiled metadata factory would be still useful. For example if you have many instances with pre-warmed-up cache, you could not need to fetch metadata at all ~  Or even with multiple FPM processes on a process you could have metadata about a class A and B (transformation from A to B with AutoMapper) and on another process you want metadata about A and C (because you want to transform A to C) then you would not required A metadata fetching because it is already compiled.

TL;DR It would be nice to keep that feature, I did not even know it. I think we should talk more about it !

**Comment by fbourigault** [comment]
I tried to do some up to date benchmark.

With the following benchmark code (loading 500k times as simple class metadata), I get a 75% improvement: https://blackfire.io/profiles/compare/986977f8-c435-437f-b920-1d64f594df1f/graph

So we are talking of a gain of 4 μs per entity if actual cache is warm.

I doubt it worth it as someone looking for such micro optimization should probably give a try at Automapper.

What do you think?

```php
<?php

use App\Person;
use Blackfire\Client;
use Blackfire\Profile\Configuration;
use Symfony\Component\Cache\Adapter\ArrayAdapter;
use Symfony\Component\Console\Application;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Serializer\Mapping\Factory\CacheClassMetadataFactory;
use Symfony\Component\Serializer\Mapping\Factory\ClassMetadataFactory;
use Symfony\Component\Serializer\Mapping\Factory\ClassMetadataFactoryCompiler;
use Symfony\Component\Serializer\Mapping\Factory\ClassMetadataFactoryInterface;
use Symfony\Component\Serializer\Mapping\Factory\CompiledClassMetadataFactory;
use Symfony\Component\Serializer\Mapping\Loader\AttributeLoader;

require __DIR__ . '/vendor/autoload.php';

$application = new Application('ClassMetadataFactory benchmark');
$command = new class extends Command {

    private Client $blackfire;

    public function __construct()
    {
        parent::__construct('benchmark');

        $this->blackfire = new Client();
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $classMetadataFactory = new ClassMetadataFactory(new AttributeLoader());
        $cacheClassMetadataFactory = new CacheClassMetadataFactory($classMetadataFactory, new ArrayAdapter());

        $classMetadataFactoryCompiler = new ClassMetadataFactoryCompiler();
        file_put_contents(__DIR__ . '/metadata.php', $classMetadataFactoryCompiler->compile([
            $cacheClassMetadataFactory->getMetadataFor(Person::class),
        ]));

        $compiledClassMetadataFactory = new CompiledClassMetadataFactory(__DIR__ . '/metadata.php', $classMetadataFactory);

        $compiledClassMetadataFactoryWithCache = new CompiledClassMetadataFactory(__DIR__ . '/metadata.php', $cacheClassMetadataFactory);

        $this->benchmark('ClassMetadataFactory', $classMetadataFactory);
        $this->benchmark('CacheClassMetadataFactory', $cacheClassMetadataFactory);
        $this->benchmark('CompiledClassMetadataFactory', $compiledClassMetadataFactory);
        $this->benchmark('CompiledClassMetadataFactory with cache', $compiledClassMetadataFactoryWithCache);

        return Command::SUCCESS;
    }

    private function benchmark(string $title, ClassMetadataFactoryInterface $classMetadataFactory): void
    {
        $configuration = new Configuration();
        $configuration->setTitle($title);

        $probe = $this->blackfire->createProbe($configuration);
        for ($i = 0; $i < 500000; $i++) {
            $classMetadataFactory->getMetadataFor(Person::class);
        }
        $this->blackfire->endProbe($probe);
    }
};

$application->add($command);

$application->setDefaultCommand($command->getName(), true);
$application->run();
```

</detail>

**Comment by mtarld** [comment]
Agreed with @fbourigault here, I don't think it worth keeping that code for such micro optimization.
Therefore, I think we might deprecate `CompiledClassMetadataFactory` and related.
